src/module_4/Module_4.md

```diff
@@ -0,0 +1,1998 @@
+    metrics = {}
+    for model_name, preds in predictions.items():
+        precision, recall, _ = precision_recall_curve(y_true, preds)
+        fpr, tpr, _ = roc_curve(y_true, preds)
+        auc_score = auc(fpr, tpr)
+        ap_score = average_precision_score(y_true, preds)
+        
+        metrics[model_name] = {
+            'precision': precision,
+            'recall': recall,
+            'fpr': fpr,
+            'tpr': tpr,
+            'auc': auc_score,
+            'ap': ap_score
+        }
+    
+    return metrics
+
+```
+
+
+```python
+def plot_metrics(metrics):
+    # Plot Precision-Recall curves
+    plt.figure(figsize=(14, 6))
+    plt.subplot(1, 2, 1)
+    for model_name, metric in metrics.items():
+        plt.plot(metric['recall'], metric['precision'], label=f'{model_name} (AP={metric["ap"]:.2f})')
+    plt.xlabel('Recall')
+    plt.ylabel('Precision')
+    plt.title('Precision-Recall Curve')
+    plt.legend()
+
+    # Plot ROC curves
+    plt.subplot(1, 2, 2)
+    for model_name, metric in metrics.items():
+        plt.plot(metric['fpr'], metric['tpr'], label=f'{model_name} (AUC={metric["auc"]:.2f})')
+    plt.xlabel('False Positive Rate')
+    plt.ylabel('True Positive Rate')
+    plt.title('ROC Curve')
+    plt.legend()
+
+    plt.tight_layout()
+    plt.show()
+
+```
+
+
+```python
+# Example usage with your predictions
+predictions = {
+    'Logistic Regression': lr_predictions,
+    'Random Forest': rf_predictions,
+    'Gradient Boosting': gb_predictions
+}
+
+metrics = compute_metrics(Y_test, predictions)
+plot_metrics(metrics)
+```
+
+
+    
+![png](Module_4_files/Module_4_58_0.png)
+    
+
+
+
+    The Kernel crashed while executing code in the current cell or a previous cell. 
+
+
+    Please review the code in the cell(s) to identify a possible cause of the failure. 
+
+
+    Click <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. 
+
+
+    View Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details.
+
+
+We can osberve that in the Precision-Recall Curve, the Logistic Regression has the highest average precision (AP),compared to Random Forest and Gradient Boosting. Furthermore, Random Forest has a moderate AP, while Gradient Boosting has the lowest AP, indicating weaker performance in distinguishing the positive class.
+
+On the other hand, in the ROC Curve, Logistic Regression and Gradient Boosting show similar performance with AUCs of 0.81 and 0.80, respectively. Random Forest performs worse with an AUC of 0.73, indicating a higher rate of false positives and/or false negatives compared to the other two models.
+
+In conclusion, Logistic Regression is the best performer in terms of both Precision-Recall and ROC metrics; Gradient Boosting has comparable ROC performance to Logistic Regression but performs worse in the Precision-Recall metric; and, finally, Random Forest has the lowest performance in both Precision-Recall and ROC metrics.
```

Comment:
It looks odd the difference between val and test. Either there is heavy data shift or bug. This is worth investigating